### PR TITLE
[clang][analyzer] Support `fputs` in the StreamChecker

### PR DESCRIPTION
**Pull request by benshi001**

### CODE REVIEW

**Review by balazske — CHANGES_REQUESTED** [review]

File: clang/test/Analysis/stream-error.c [hunk 1]
```diff
@@ -141,6 +141,24 @@ void error_fputc(void) {
...
+  } else {
+    clang_analyzer_eval(Ret == EOF); // expected-warning {{TRUE}}
+    clang_analyzer_eval(ferror(F));  // expected-warning {{TRUE}}
+    clang_analyzer_eval(feof(F));    // expected-warning {{FALSE}}
+    fputs("QWD", F);                 // expected-warning {{might be 'indeterminate'}}
+  }
+  fclose(F);
+  fputs("ABC", F);                   // expected-warning {{Stream might be already closed}}
+}
+
```
Comment (balazske) [inline clang/test/Analysis/stream-error.c:161]: One type of test is missing, where the stream is in EOF state (`StreamTesterChecker_make_feof_stream(F)` can be used to get such a stream `F`) and a `fputc` or `fputs` is called. Then no error and no warning should be generated, and the state after the call should be non-error and non-EOF.

File: clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp [hunk 2]
```diff
@@ -824,20 +817,76 @@ void StreamChecker::evalFgetcFputc(const FnDescription *Desc,
...
+        StateNotFailed->set<StreamMap>(StreamSym, StreamState::getOpened(Desc));
+    C.addTransition(StateNotFailed);
+  }
+  // Generddate a transition for the success state of `fputs`.
+  else {
+    NonLoc RetVal = makeRetVal(C, CE).castAs<NonLoc>();
+    ProgramStateRef StateNotFailed =
+        State->BindExpr(CE, C.getLocationContext(), RetVal);
+    SValBuilder &SVB = C.getSValBuilder();
+    auto &ASTC = C.getASTContext();
```
[the remark this answers is not in the document]
Comment (balazske) [reply clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp:866]: Line 866 still contains `auto`, usually `auto` is not used at such places (https://llvm.org/docs/CodingStandards.html#id29).

**Comment by balazske** [comment]
I prefer to have the new test code with `StreamTesterChecker_make_feof_stream` in a new test function (with appropriate name like `write_after_eof_is_allowed`), to improve code maintainability (`fputc` and `fputs` can be in the same function).

In github PR's usually force push is not the good way to update the code (it makes difficult to follow what was updated compared to the previous code). It is better to add new commits for the changes. The final "squash and merge" will merge all into one commit.

**Comment by benshi001** [comment]
> I prefer to have the new test code with `StreamTesterChecker_make_feof_stream` in a new test function (with appropriate name like `write_after_eof_is_allowed`), to improve code maintainability (`fputc` and `fputs` can be in the same function).
> 
> In github PR's usually force push is not the good way to update the code (it makes difficult to follow what was updated compared to the previous code). It is better to add new commits for the changes. The final "squash and merge" will merge all into one commit.

I have added a new test `write_after_eof_is_allowed` as you mentioned. Thanks.

**Review by balazske — APPROVED** [review]

**Pull request merged**